python/ray/tune/examples/dcgan_mnist_pbt/dcgan_mnist_pbt.py

```diff
@@ -0,0 +1,364 @@
+        py = np.mean(part, axis=0)
+        scores = []
+        for i in range(part.shape[0]):
+            pyx = part[i, :]
+            scores.append(entropy(pyx, py))
+        split_scores.append(np.exp(np.mean(scores)))
+
+    return np.mean(split_scores), np.std(split_scores)
+# __INCEPTION_SCORE_end__
+
+
+def train(netD, netG, optimG, optimD, criterion, dataloader, iteration):
+    real_label = 1
+    fake_label = 0
+
+    for i, data in enumerate(dataloader, 0):
+        if i >= 5:
+            break
+
+        netD.zero_grad()
+        real_cpu = data[0].to(device)
+        b_size = real_cpu.size(0)
+        label = torch.full((b_size,), real_label, device=device)
+        output = netD(real_cpu).view(-1)
+        errD_real = criterion(output, label)
+        errD_real.backward()
+        D_x = output.mean().item()
+
+        noise = torch.randn(b_size, nz, 1, 1, device=device)
+        fake = netG(noise)
+        label.fill_(fake_label)
+        output = netD(fake.detach()).view(-1)
+        errD_fake = criterion(output, label)
+        errD_fake.backward()
+        D_G_z1 = output.mean().item()
+        errD = errD_real + errD_fake
+        optimD.step()
+
+        netG.zero_grad()
+        label.fill_(real_label)
+        output = netD(fake).view(-1)
+        errG = criterion(output, label)
+        errG.backward()
+        D_G_z2 = output.mean().item()
+        optimG.step()
+
+        is_score, is_std = inception_score(fake)
+
+        # Output training stats
+        if iteration % 10 == 0:
+            print('[%d/%d]\tLoss_D: %.4f\tLoss_G: %.4f\tD(x): %.4f\tD(G(z))'
+                  ': %.4f / %.4f \tInception score: %.4f'
+                  % (iteration, len(dataloader),
+                     errD.item(), errG.item(), D_x, D_G_z1, D_G_z2, is_score))
+
+    return errG.item(), errD.item(), is_score
+
+
+# __Trainable_begin__
+class PytorchTrainable(tune.Trainable):
+    def _setup(self, config):
+        self.netD = Discriminator().to(device)
+        self.netD.apply(weights_init)
+        self.netG = Generator().to(device)
+        self.netG.apply(weights_init)
+        self.criterion = nn.BCELoss()
+        self.optimizerD = optim.Adam(self.netD.parameters(),
+                                     lr=config.get("lr", 0.01),
+                                     betas=(beta1, 0.999))
+        self.optimizerG = optim.Adam(self.netG.parameters(),
+                                     lr=config.get("lr", 0.01),
+                                     betas=(beta1, 0.999))
+        self.dataloader = get_data_loader()
+
+    def _train(self):
+        lossG, lossD, is_score = train(self.netD, self.netG, self.optimizerG,
+                                       self.optimizerD, self.criterion,
+                                       self.dataloader, self._iteration)
+        return {"lossg": lossG, "lossd": lossD, "is_score": is_score}
+
+    def _save(self, checkpoint_dir):
```

Comment:
Ah, one thing that we did not do correctly in other examples is that we need to save the state_dict of the optimizer.

``self.optimizerG.state_dict()`` and ``self.optimizerD.state_dict()``

Reply:
Sure. Updated.